[quant][graphmode] Skip inserting duplicate observers

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #25974 [quant][graphmode] use whitelist for selecting observed values
* #25625 [quant][graphmode] Fold quantize op into module
* #25624 [quant][graphmode] Add fusion for quantized linear
* #25623 Port fuse_linear from pytorch/tvm
* #25505 [quant][graphmode] Support quantizing any methods called
* **#25504 [quant][graphmode] Skip inserting duplicate observers**

Summary:
Skip inserting duplicate observers for values observed
in forward method of a child module or other methods in
the current module.

Test Plan:
python test/test_jit.py -- 'TestJit.insert_observers'
python test/test_jit.py -- 'TestJit.insert_observers_child_qconfig'
python test/test_jit.py -- 'TestJit.insert_observers_skip_values'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208888](https://our.internmc.facebook.com/intern/diff/D17208888)

jerryzh168 committed:
[quant][graphmode] Skip inserting duplicate observers

Summary:
Skip inserting duplicate observers for values observed
in forward method of a child module or other methods in
the current module.

Test Plan:
python test/test_jit.py -- 'TestJit.insert_observers'
python test/test_jit.py -- 'TestJit.insert_observers_child_qconfig'
python test/test_jit.py -- 'TestJit.insert_observers_skip_values'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] Skip inserting duplicate observers"


Summary:
Skip inserting duplicate observers for values observed
in forward method of a child module or other methods in
the current module.

Test Plan:
python test/test_jit.py -- 'TestJit.insert_observers'
python test/test_jit.py -- 'TestJit.insert_observers_child_qconfig'
python test/test_jit.py -- 'TestJit.insert_observers_skip_values'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] Skip inserting duplicate observers"


Summary:
Skip inserting duplicate observers for values observed
in forward method of a child module or other methods in
the current module.

Test Plan:
python test/test_jit.py -- 'TestJit.insert_observers'
python test/test_jit.py -- 'TestJit.insert_observers_child_qconfig'
python test/test_jit.py -- 'TestJit.insert_observers_skip_values'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] Skip inserting duplicate observers"


Summary:
Skip inserting duplicate observers for values observed
in forward method of a child module or other methods in
the current module.

Test Plan:
python test/test_jit.py -- 'TestJit.insert_observers'
python test/test_jit.py -- 'TestJit.insert_observers_child_qconfig'
python test/test_jit.py -- 'TestJit.insert_observers_skip_values'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] Skip inserting duplicate observers"


Summary:
Skip inserting duplicate observers for values observed
in forward method of a child module or other methods in
the current module.

Test Plan:
python test/test_jit.py -- 'TestJit.insert_observers'
python test/test_jit.py -- 'TestJit.insert_observers_child_qconfig'
python test/test_jit.py -- 'TestJit.insert_observers_skip_values'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 commented:
I found one case that is not handled by the current design, if we have a module that is used multiple times, e.g. self.relu, then we actually need to either insert observers in the parent module or copy the reused module in parent: https://github.com/pytorch/vision/blob/master/torchvision/models/resnet.py#L71, I prefer the second choice since I feel for first choice, detecting a module is used multiple times might be hard and it requires us to implement two set of logic of inserting observers, if we go with the second choice, I think we need to do another pass to make sure all the submodules are used once, i.e. copy the modules that is used multiple times.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -157,47 +219,38 @@ Node* insertObserver(
-}
-
-void getQConfigMapHelper(
+void getQConfigMap(
```
The name is now misleading, because `get` usually returns something.

jerryzh168 replied:
changing to fillQConfigMap

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -64,6 +64,68 @@ graph(%self, %input):
+  return false;
+}
+
+bool valueObservedInOtherMethod(
```
Nit: "Another" instead of "Other" :p

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -64,6 +64,68 @@ graph(%self, %input):
+}
+
+bool valueObserved(
+    Node* n,
```
Nit: the function is called "valueObserved" but takes `Node` instead of a `Value`. It would probably be better to take value as an argument and get the corresponding node inside.

Also, the meaning of the other parameters is unclear. Even though it's not a public API, it's worth to add a comment about it.

jerryzh168 replied:
oh, it's node because we only have access to node here:     
`if (valueObserved(u.user, self, module, child_module_set)) {`, but I can change the function name.

ZolotukhinM replied:
What are we observing, a node or a value? I think observing values is the correct abstraction here, and in this case I would rather suggest extracting the value from `Use` (`u.offset` will give you an the index of the value in `u.user.inputs()`), rather than rename the function to `nodeObserved` - what would it even mean if the node had several arguments and one of them was observed and the other wasn't?

jerryzh168 replied:
oh, I'm not changing the name to `nodeObserved`, I changed it to `callMethodOfModuleWithSameQConfig`

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -64,6 +64,68 @@ graph(%self, %input):
+    Value* self,
+    const script::Module& module,
+    const ModulePtrSet& child_module_set) {
+  if (n->kind() == prim::CallMethod) {
```
Nit: we could use early exit here.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -64,6 +64,68 @@ graph(%self, %input):
+    if (module_instance == self) {
+      // Calling a method of current module
+      return true;
+    } else {
```
Nit: `else` is unnecessary.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -580,8 +629,20 @@ TORCH_API void InsertObservers(
-  InsertObserversImpl(module, method_name, module_qconfig_map);
+  ModuleQConfigMap module_qconfig_map;
+  // Set of child modules that has the same qconfig
+  // as parent module, used for avoiding inserting
```
Why can't we just add these children to the map we're generating (so that `qmap.at(child)` == `qmap.at(parent)`)? Having an extra set seems unnecessary and also confusing, I had to reread the code and the comment several times and I'm still not sure I understood it correctly.

jerryzh168 replied:
the reason is we can't compare QConfig, it's a tuple of two script::Module and we don't have a way to compare script::Modules I think. That's why I record all the child modules that inherit the qconfig of parent in the place when we are filling the QConfigMap.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -580,8 +629,20 @@ TORCH_API void InsertObservers(
+  ModulePtrSet child_module_set;
+  getQConfigMap(module,
+                qconfig_dict,
+                "",
```
These parameters should probably have default values so that this invocation looks simple. Right now the reader is unnecessarily exposed to internal details of the function implementation.

jerryzh168 replied:
I can make these two default

ZolotukhinM reviewed (changes requested):
I left some initial comments, but I need to think more about how we want to handle values observed inside and outside of a call.

jerryzh168 committed:
Update on "[quant][graphmode] Skip inserting duplicate observers"


Summary:
Skip inserting duplicate observers for values observed
in forward method of a child module or other methods in
the current module.

Test Plan:
python test/test_jit.py -- 'TestJit.insert_observers'
python test/test_jit.py -- 'TestJit.insert_observers_child_qconfig'
python test/test_jit.py -- 'TestJit.insert_observers_skip_values'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208888](https://our.internmc.facebook.com/intern/diff/D17208888)

nishantpdce commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -490,8 +583,7 @@ void QuantizeHelper::quantizeTensor(Value* v, bool insert_after) {
     Value* v) {
-  if (v->node()->kind() == prim::CallMethod &&
-      v->node()->s(attr::name) == "forward") {
+  if (v->node()->kind() == prim::CallMethod) {
```
why is check for "forward" removed? Is it because this will be called recursively and child may not have "forward" ?

jerryzh168 replied:
yes, we are quantizing all the method that is called

nishantpdce commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -217,9 +299,16 @@ void InsertObserversImpl(
+    if (v->type()->isSubtypeOf(TensorType::get()) &&
+        values_to_skip.count(v) == 0 &&
+        !valueObservedInAnotherMethod(v, self, module, child_module_set)) {
+      if (module_qconfig_map.count(module.module_object()) <= 0) {
```
Very minor nit - can this be < 0 ever?

jerryzh168 replied:
ok will change to 0

jerryzh168 committed:
Update on "[quant][graphmode] Skip inserting duplicate observers"


Summary:
Skip inserting duplicate observers for values observed
in forward method of a child module or other methods in
the current module.

Test Plan:
python test/test_jit.py -- 'TestJit.insert_observers'
python test/test_jit.py -- 'TestJit.insert_observers_child_qconfig'
python test/test_jit.py -- 'TestJit.insert_observers_skip_values'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208888](https://our.internmc.facebook.com/intern/diff/D17208888)

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -56,12 +57,100 @@ graph(%self, %input):
+
+// Propagate current values_to_skip through module hierarchy,
+// adding values to skip in child method graph
+void inputOutputValuesToSkipObserver(
```
FYI, this part of code will be deleted in the later diff, so no need to spend too much time reviewing this part

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -56,12 +57,100 @@ graph(%self, %input):
+// 1. calling a method of self
+// 2. calling a method of child module that has the same
+// qconfig as the current module
+bool callMethodOfModuleWithSameQConfig(
```
same for this one, this is not needed in later PR

jerryzh168 committed:
Update on "[quant][graphmode] Skip inserting duplicate observers"


Summary:
Skip inserting duplicate observers for values observed
in forward method of a child module or other methods in
the current module.

Test Plan:
python test/test_jit.py -- 'TestJit.insert_observers'
python test/test_jit.py -- 'TestJit.insert_observers_child_qconfig'
python test/test_jit.py -- 'TestJit.insert_observers_skip_values'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208888](https://our.internmc.facebook.com/intern/diff/D17208888)

jerryzh168 committed:
Update on "[quant][graphmode] Skip inserting duplicate observers"


Summary:
Skip inserting duplicate observers for values observed
in forward method of a child module or other methods in
the current module.

Test Plan:
python test/test_jit.py -- 'TestJit.insert_observers'
python test/test_jit.py -- 'TestJit.insert_observers_child_qconfig'
python test/test_jit.py -- 'TestJit.insert_observers_skip_values'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208888](https://our.internmc.facebook.com/intern/diff/D17208888)

jerryzh168 committed:
Update on "[quant][graphmode] Skip inserting duplicate observers"


Summary:
Skip inserting duplicate observers for values observed
in forward method of a child module or other methods in
the current module.

Test Plan:
python test/test_jit.py -- 'TestJit.insert_observers'
python test/test_jit.py -- 'TestJit.insert_observers_child_qconfig'
python test/test_jit.py -- 'TestJit.insert_observers_skip_values'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208888](https://our.internmc.facebook.com/intern/diff/D17208888)

jerryzh168 committed:
Update on "[quant][graphmode] Skip inserting duplicate observers"


Summary:
Skip inserting duplicate observers for values observed
in forward method of a child module or other methods in
the current module.

Test Plan:
python test/test_jit.py -- 'TestJit.insert_observers'
python test/test_jit.py -- 'TestJit.insert_observers_child_qconfig'
python test/test_jit.py -- 'TestJit.insert_observers_skip_values'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208888](https://our.internmc.facebook.com/intern/diff/D17208888)

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -217,9 +207,15 @@ void InsertObserversImpl(
-    if (v->type()->isSubtypeOf(TensorType::get())) {
+    if (v->type()->isSubtypeOf(TensorType::get()) &&
+        values_to_skip.count(v) == 0) {
+      if (module_qconfig_map.count(module.module_object()) == 0) {
```
This check should be moved to the beginning of the function.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -580,8 +568,20 @@ TORCH_API void InsertObservers(
-  auto module_qconfig_map = getQConfigMap(module, qconfig_dict);
-  InsertObserversImpl(module, method_name, module_qconfig_map);
+  ModuleQConfigMap module_qconfig_map;
+  // Set of child modules that has the same qconfig
```
This comment seems to be outdated - we no longer have childset which I believe it refers to.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -28,13 +29,14 @@ void findValuesInPattern(
 }
 
-std::unordered_set<Value*> valuesToSkipObserver(
+void intermediateValuesToSkipObserver(
```
Nit: might be better to use a verb for the function name.

ZolotukhinM reviewed (approved):
Looks good, I only have minor comments! Thanks for following up on my earlier feedback!

jerryzh168 committed:
Update on "[quant][graphmode] Skip inserting duplicate observers"


Summary:
Skip inserting duplicate observers for values observed
in forward method of a child module or other methods in
the current module.

Test Plan:
python test/test_jit.py -- 'TestJit.insert_observers'
python test/test_jit.py -- 'TestJit.insert_observers_child_qconfig'
python test/test_jit.py -- 'TestJit.insert_observers_skip_values'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208888](https://our.internmc.facebook.com/intern/diff/D17208888)

jerryzh168 committed:
Update on "[quant][graphmode] Skip inserting duplicate observers"


Summary:
Skip inserting duplicate observers for values observed
in forward method of a child module or other methods in
the current module.

Test Plan:
python test/test_jit.py -- 'TestJit.insert_observers'
python test/test_jit.py -- 'TestJit.insert_observers_child_qconfig'
python test/test_jit.py -- 'TestJit.insert_observers_skip_values'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208888](https://our.internmc.facebook.com/intern/diff/D17208888)
